src/cheri_cap_common.sail

```diff
@@ -236,41 +236,48 @@ function capToMemBits(cap) : Capability -> CapBits =
+  let base : CapLenBits = a_top @ c.B @ zeros(E) in
+  let top  : CapLenBits = a_top @ c.T @ zeros(E) in
+    (base, top)
+
+function getCapLegalised(c) : Capability -> Capability =
+  let (base, top) = getCapBoundsBitsUncorrected(c) in
+  /* Check that base is within the address space */
+  if unsigned(c.E) <= cap_max_E
+   & (base <_u (to_bits(cap_len_width, 1) << cap_addr_width)) then
+    c
+  else
+    /* Legalise the capability by zeroing the ie bit, implicitly setting E to 0. */
+    encCapabilityToCapability(false, {capToEncCap(c) with internal_E = bitzero})
+
+function getCapBoundsBits(c) : Capability -> (CapAddrBits, CapLenBits) =
+  let c_legal = getCapLegalised(c) in
+  let E = unsigned(c_legal.E) in
+  let a : CapAddrBits = c_legal.address in
+  let (base, top) = getCapBoundsBitsUncorrected(c_legal) in
   /* Extract bits we need to make the top correction and calculate representable limit */
   let a3 = truncate(a >> (E + cap_mantissa_width - 3), 3) in
-  let B3 = truncateLSB(c.B, 3) in
-  let T3 = truncateLSB(c.T, 3) in
+  let B3 = truncateLSB(c_legal.B, 3) in
+  let T3 = truncateLSB(c_legal.T, 3) in
   let R3 = B3 - 0b001 in /* wraps */
   /* Do address, base and top lie in the R aligned region above the one containing R? */
   let aHi = if a3 <_u R3 then 1 else 0 in
   let bHi = if B3 <_u R3 then 1 else 0 in
   let tHi = if T3 <_u R3 then 1 else 0 in
   /* Compute region corrections for top and base relative to a */
   let correction_base = bHi - aHi in
-  let correction_top  = tHi - aHi in
-  let a_top = (a >> (E + cap_mantissa_width)) in {
-    base : CapLenBits = truncate((a_top + correction_base) @ c.B @ zeros(E), cap_len_width);
-    top  : CapLenBits = truncate((a_top + correction_top)  @ c.T @ zeros(E), cap_len_width);
-    /* If the base and top are more than an address space away from each other,
-       invert the MSB of top.  This corrects for errors that happen when the
-       representable space wraps the address space. */
-    let base2 : bits(2) = 0b0 @ [base[cap_addr_width - 1]];
-    let top2  : bits(2) = top[cap_addr_width .. cap_addr_width - 1];
-    if (E < (cap_max_E - 1)) & (unsigned(top2 - base2) > 1) then {
-      top[cap_addr_width] = ~(top[cap_addr_width]);
-    };
-    /* TODO The following top MSB correction was thought to be equivalent
-       to the one above, but differs at least in some cases where E >= cap_max_E - 1
-    if (base[cap_addr_width] == bitone) then {
-       top[cap_addr_width] = if (aHi == 1) & (tHi == 1) then bitone else bitzero;
-    };
-    */
-    (base[(cap_addr_width - 1)..0], top)
+  let correction_top  = tHi - aHi in {
+    base_corrected : CapLenBits = base + (to_bits(cap_len_width, correction_base) << (cap_mantissa_width + E));
+    top_corrected : CapLenBits = top + (to_bits(cap_len_width, correction_top) << (cap_mantissa_width + E));
+    /* Interpret the top relative to the address space of the base.
```

Comment:
Could you expand on this comment, e.g. explaining why this prevents base > top?

Reply:
Hmm, I don't think this alone prevents that. It's more ensuring that the top hasn't wrapped around to be in an entire address space (or more) below the base? I must admit, my memory is slightly hazy.

Reply:
Yeah it's been a while since I looked at this which is why I wasn't sure why we can now assert that base <= top. Adding some comments explaining where this is guaranteed would be helpful for future readers I believe.

Reply:
Right, agreed. I guess "it's been proved" is something, but it's also nice to see how it's been constructed to make this true.